docs/source/gbd2019_models/hemoglobin_anemia_and_iron_deficiency/hemoglobin/index.rst

```diff
@@ -6,4 +6,290 @@ Hemoglobin Model
+
+.. code-block:: R
+
+  # define constants
+  XMAX = 220
+  EULERS_CONSTANT = 0.57721566490153286060651209008240243104215933593992
+  w = c(0.4,0.6)
+
+  # import standard R functions for the gamma distributions (pgamma and rgamma)
+  pacman::p_load(data.table,actuar)
+
+  # function to calculate gamma distribution parameters from mean and variance
+  gamma_mv2p = function(mn, vr){
+    list(shape = mn^2/vr,rate = mn/vr)}
+
+  # function to calculate mirror gumbel distribution parameters from mean and variance 
+  mgumbel_mv2p = function(mn, vr){
+    list(
+      alpha = XMAX - mn - EULERS_CONSTANT*sqrt(vr)*sqrt(6)/pi,
+      scale = sqrt(vr)*sqrt(6)/pi)}
+  # function to randomly sample n times from mirror gumbel distribution
+  rmgumbel = function(n, alpha, scale){
+    mn = alpha + scale*EULERS_CONSTANT
+    rgumbel(n, alpha+XMAX-(2*mn), scale)}
+
+  # function to sample from ensemble hemoglobin distribution using functions defined above
+    # n = number of samples to take
+    # mn = mean hemoglobin concentration
+    # vr = hemoglobin distribution variance (standard deviation squared)
+    # w = list of ensemble distribution weights c(gamma_weight, mirror_gumbel_weight)
+  sample_from_hemoglobin_distribution <- function(n, mn, vr, w){
+
+    ##parameters
+    params_gamma = gamma_mv2p(mn, vr)
+    params_mgumbel = mgumbel_mv2p(mn, vr)
+
+    ##weighting
+    sample_list = sum(
+      w[1] * rgamma(n, params_gamma$shape, params_gamma$rate), 
+      w[2] * rmgumbel(n, params_mgumbel$alpha, params_mgumbel$scale)
+    )
+    sample_list
+  }
+
+  # function to calculate area under curve below threshold q for mirror gumbel distribution
+  pmgumbel = function(q, alpha, scale, lower.tail){ 
+    #NOTE: with mirroring, take the other tail
+    pgumbel(XMAX-q, alpha, scale, lower.tail=ifelse(lower.tail,FALSE,TRUE))}
+
+  # function to calculate area under curve of hemoglobin ensemble distribution using the functions defined above
+    # q = hemoglobin threshold
+    # mn = mean hemoglobin concentration
+    # vr = hemoglobin variance (standard deviation squared)
+    # w = list of ensemble distribution weights c(gamma_weight, mirror_gumbel_weight)
+  ens_mv2prev <- function(q, mn, vr, w){
+    x = q
+
+    ##parameters
+    params_gamma = gamma_mv2p(mn, vr)
+    params_mgumbel = mgumbel_mv2p(mn, vr)
+
+    ##weighting
+    prev = sum(
+      w[1] * pgamma(x, params_gamma$shape, params_gamma$rate), 
+      w[2] * pmgumbel(x, params_mgumbel$alpha, params_mgumbel$scale, lower.tail=T))
+    prev}
+
+.. note::
+
+	While not explicitly enforced by the code above, all hemoglobin values should be withn the bounds of 0 and 220 (XMAX). The probability of sampling a value outside of these bounds is small, but if it occurs, the value should be resampled until it falls within the bounds of 0 and 220.
```

Comment:
I agree that we should make sure all Hb values are >= 0 (do we want to resample, or just [clip](https://numpy.org/doc/stable/reference/generated/numpy.clip.html) the values to a small positive number?), but do we actually need to ensure that they're <= 220?  Is this assumption enforced anywhere in the GBD code?  My impression from a cursory look through the R code is that XMAX is really just an "approximate" maximum that is used to parameterize the distributions, but is not necessarily a hard maximum (though I could be wrong).

In particular, in [Kjjell's notebook](https://github.com/ihmeuw/vivarium_gates_lsff/blob/master/tests/lsff_iron_exposure.ipynb), sampling 10,000 draws from the ensemble gave values as high as 243. Do we actually need Vivarium to use rejection sampling to resample until all draws are <=220 (another option would be to just `np.clip` all the larger values to exactly 220), or is it fine to leave the large values?

Reply:
I agree that we don't need to keep values under 220 and added that values could be clipped to 1 instead of being resampled.